narwhal/primary/src/core.rs

```diff
@@ -107,14 +111,20 @@ impl Core {
         gc_depth: Round,
         rx_committee: watch::Receiver<ReconfigureNotification>,
         rx_primaries: Receiver<PrimaryMessage>,
-        rx_header_waiter: Receiver<Header>,
-        rx_certificate_waiter: Receiver<Certificate>,
+        rx_headers_loopback: Receiver<Header>,
+        rx_certificates_loopback: Receiver<CertificateLoopbackMessage>,
         rx_proposer: Receiver<Header>,
         tx_consensus: Sender<Certificate>,
         tx_proposer: Sender<(Vec<Certificate>, Round, Epoch)>,
         metrics: Arc<PrimaryMetrics>,
         primary_network: P2pNetwork,
     ) -> JoinHandle<()> {
+        let last_round = certificate_store
+            .last_round()
+            .unwrap()
+            .first()
+            .map(|c| c.round())
+            .unwrap_or(0);
```

Comment:
I would suggest to do that as part of the `recover()` method which is already reading the last round certificates and is exclusively dealing with the component's recovery.

Reply:
Good point!